### PR TITLE
[rpcwallet] Clamp walletpassphrase value at 100M seconds

### PR DESCRIPTION
**Pull request by sdaftuar**
Larger values seem to trigger a bug on macos+libevent (resulting in the rpc server stopping).

### CODE REVIEW

**Comment by sdaftuar** [comment]
This is a quick fix for #12375.  Not sure if further reducing this value as I've proposed here is acceptable, or if we should be trying to solve this problem at a different level?

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2349,8 +2349,7 @@ UniValue walletpassphrase(const JSONRPCRequest& request)
             "This is needed prior to performing transactions related to private keys such as sending bitcoins\n"
             "\nArguments:\n"
             "1. \"passphrase\"     (string, required) The wallet passphrase\n"
-            "2. timeout            (numeric, required) The time to keep the decryption key in seconds. Limited to at most 1073741824 (2^30) seconds.\n"
-            "                                          Any value greater than 1073741824 seconds will be set to 1073741824 seconds.\n"
+            "2. timeout            (numeric, required) The time to keep the decryption key in seconds; capped at 100000000.\n"
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: nit: Could add a comment that the constant is about 3 years.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -2383,9 +2382,10 @@ UniValue walletpassphrase(const JSONRPCRequest& request)
     if (nSleepTime < 0) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Timeout cannot be negative.");
     }
-    // Clamp timeout to 2^30 seconds
-    if (nSleepTime > (int64_t)1 << 30) {
-        nSleepTime = (int64_t)1 << 30;
+    // Clamp timeout
+    int64_t max_sleep_time = 100000000; // larger values trigger a macos/libevent bug?
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: I think we currently use `constexpr` (without `static`) and UPPER_CASE for compile time constants.

File: test/functional/wallet_encryption.py [hunk 3]
```diff
@@ -64,14 +64,15 @@ def run_test(self):
         assert_raises_rpc_error(-8, "Timeout cannot be negative.", self.nodes[0].walletpassphrase, passphrase2, -10)
         # Check the timeout
         # Check a time less than the limit
-        expected_time = int(time.time()) + (1 << 30) - 600
-        self.nodes[0].walletpassphrase(passphrase2, (1 << 30) - 600)
+        max_value = 100000000
```
Comment (MarcoFalke) [inline test/functional/wallet_encryption.py]: nit: Constants should be UPPER_CASE

File: test/functional/wallet_encryption.py [hunk 4]
```diff
@@ -64,14 +64,15 @@ def run_test(self):
...
+        expected_time = int(time.time()) + max_value - 600
+        self.nodes[0].walletpassphrase(passphrase2, max_value - 600)
         actual_time = self.nodes[0].getwalletinfo()['unlocked_until']
         assert_greater_than_or_equal(actual_time, expected_time)
         assert_greater_than(expected_time + 5, actual_time) # 5 second buffer
         # Check a time greater than the limit
-        expected_time = int(time.time()) + (1 << 30) - 1
-        self.nodes[0].walletpassphrase(passphrase2, (1 << 33))
+        expected_time = int(time.time()) + max_value - 1
+        self.nodes[0].walletpassphrase(passphrase2, max_value + 1)
```
Comment (MarcoFalke) [inline test/functional/wallet_encryption.py]: Might want to be off by more than `1`, since we have a "buffer" of 5? Maybe try off by 1000 or something?

**Review by maflcko — COMMENTED** [review]
utACK, just a ton of style nits.

Also, since you are touching the `walletpassphrase` code anyway, you might want to change the `boost::bind` into `std::bind` to preempt future pull request from doing that.

**Comment by sdaftuar** [comment]
Thanks @MarcoFalke, nits fixed in latest commit.

**Comment by promag** [comment]
utACK, also add release note?

**Comment by sipa** [comment]
Is the fact that wallets get locked after 3+ years really observable?

**Comment by laanwj** [comment]
From a principled point of view I'd still prefer to never unlock if a value higher than *defined threshold* is given, instead of clamping.
Because that's what users intend if they give a huge number like this. They don't expect the wallet to ever re-lock.
But I also agree with @sipa that no one is going to notice this in practice.

utACK - don't think this requires specific mention in the release notes.

**Comment by PierreRochard** [comment]
Tested ACK 93e179853e89e14cf248a7e36aa51ed611727ed8 (on macOS).

My view is that we should throw an error if the value provided is > MAX_SLEEP_TIME and we should have an explicit boolean `never_timeout` parameter. But in terms of fixing this macOS-related bug the PR lgtm.

**Comment by promag** [comment]
Tested ACK 93e1798.

**Comment by jonasschnelli** [comment]
utACK 93e179853e89e14cf248a7e36aa51ed611727ed8

**Comment by MarcoFalke** [comment]
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

**Comment by sdaftuar** [comment]
Squashed https://github.com/bitcoin/bitcoin/commit/93e179853e89e14cf248a7e36aa51ed611727ed8 -> https://github.com/bitcoin/bitcoin/pull/12905/commits/2b2b96cd452bd08537310ce78bee9cbf6f43c10e

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -37,6 +37,8 @@
 
 #include <univalue.h>
 
+#include <functional>
```
Comment (PierreRochard) [inline src/wallet/rpcwallet.cpp]: Why is this include needed?

**Comment by MarcoFalke** [comment]
utACK 2b2b96cd45 (Checked that it is the same code as the previous 93e179853e)

File: src/wallet/rpcwallet.cpp (hunk 5 above)
Comment (MarcoFalke) [reply src/wallet/rpcwallet.cpp]: According to http://en.cppreference.com/w/cpp/utility/functional/bind, `std::bind` is "[d]efined in header `<functional>`".

**Comment by PierreRochard** [comment]
reACK 2b2b96cd452bd08537310ce78bee9cbf6f43c10e

**Pull request merged**